### PR TITLE
1263: fs; bfd_disasm

### PR DESCRIPTION
**Pull request by Horki**
##### Checklist

- [ ] Language changes are updated in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

@fbs related to #1263 for `bfd_disasm`

### CODE REVIEW

File: src/bfd-disasm.cpp [hunk 1]
```diff
@@ -16,34 +14,17 @@
...
-  fd_ = open(path.c_str(), O_RDONLY);
-
-  if (fd_ >= 0) {
-    struct stat st;
-
-    if (fstat(fd_, &st) == 0)
-      size_ = st.st_size;
-  }
+  std::error_code ec;
+  size_ += std_filesystem::file_size(fs_, ec);
```
Comment (danobi) [inline src/bfd-disasm.cpp:20]: Can we check `ec` for errors just in case? Don't see anything on cppreference on return value if there's an error.

File: src/bfd-disasm.h [hunk 2]
```diff
@@ -1,19 +1,27 @@
...
 #include "disasm.h"
+#if __has_include(<filesystem>)
+#include <filesystem>
+namespace std_filesystem = std::filesystem;
+#elif __has_include(<experimental/filesystem>)
+#include <experimental/filesystem>
+namespace std_filesystem = std::experimental::filesystem;
+#else
+#error "neither <filesystem> nor <experimental/filesystem> are present"
+#endif
```
Comment (danobi) [inline src/bfd-disasm.h:12]: Can this be extracted out into another file, maybe `filesystem_wrapper.h`? I see another copy paste of this in `utils.cpp`. Would be great to not repeat this and also avoid using transitively included code.

File: src/bfd-disasm.cpp [hunk 3]
```diff
@@ -16,34 +14,17 @@
...
-  ssize_t len;
-
-  len = readlink(path, tpath, size - 1);
-  if (len < 0)
-    len = 0;
-
-  tpath[len] = 0;
+  std::error_code ec;
+  std_filesystem::path path{ "/proc/self/exe" };
+  return std_filesystem::read_symlink(path, ec);
```
Comment (danobi) [inline src/bfd-disasm.cpp:27]: Same as above

File: src/bfd-disasm.h (hunk 2 above)
Comment (Horki) [reply src/bfd-disasm.h:12]: @danobi sure

File: src/bfd-disasm.cpp (hunk 1 above)
Comment (Horki) [reply src/bfd-disasm.cpp:20]: @danobi sure

File: src/filesystem_wrapper.h [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include "disasm.h"
```
Comment (danobi) [inline src/filesystem_wrapper.h:3]: This shouldn't be here

File: src/bfd-disasm.cpp [hunk 5]
```diff
@@ -114,7 +89,7 @@ AlignState BfdDisasm::is_aligned(uint64_t offset, uint64_t pc)
   uint64_t size = std::min(pc + 100, size_);
   void *buf;
 
-  if (fd_ < 0)
+  if (!std_filesystem::exists(fs_))
```
Comment (danobi) [inline src/bfd-disasm.cpp:92]: If we're just going to open the fd later again I think it's better we use `open()` to avoid TOCTOU issues.

Comment (Horki) [reply src/bfd-disasm.cpp]: @danobi replaced with buffer as a unique pointer, and `ifstream` instead of C-style open, pread

File: src/filesystem_wrapper.h (hunk 4 above)
Comment (Horki) [reply src/filesystem_wrapper.h]: @fbs @danobi Hi, removed with the latest commit

File: src/bfd-disasm.cpp [hunk 6]
```diff
@@ -1,49 +1,27 @@
...
 #include <bcc/bcc_syms.h>
 #include <bfd.h>
 #include <dis-asm.h>
 
 namespace bpftrace {
 
-BfdDisasm::BfdDisasm(std::string &path) : size_(0)
+BfdDisasm::BfdDisasm(std::string &path) : fs_{ path }, size_{ 0 }
 {
-  fd_ = open(path.c_str(), O_RDONLY);
```
Comment (danobi) [inline src/bfd-disasm.cpp:21]: I strongly feel we should leave the FD based stuff in. This is an easily preventable TOCTOU bug waiting to happen that nobody enjoys debugging.

File: src/bfd-disasm.cpp (hunk 5 above)
Comment (danobi) [reply src/bfd-disasm.cpp]: I do not believe that solves the problem

Comment (Horki) [reply src/bfd-disasm.cpp]: @danobi Hi, revered to FD

File: src/bfd-disasm.h [hunk 7]
```diff
@@ -7,13 +7,13 @@ namespace bpftrace {
...
 public:
-  BfdDisasm(std::string &path);
+  BfdDisasm(std::string &);
   ~BfdDisasm();
 
   AlignState is_aligned(uint64_t offset, uint64_t pc);
 
 private:
-  int fd_ = -1;
+  int fd_;
```
Comment (danobi) [inline src/bfd-disasm.h:16]: Please revert these changes -- no point creating noise in git blame

Comment (Horki) [reply src/bfd-disasm.h]: @danobi reverted

**Comment by danobi** [comment]
Needs a rebase

**Comment by Horki** [comment]
@danobi hi, rebased

**Comment by fbs** [comment]
We try to avoid merge commits in our, it makes history confusing and can make diffs weird like happens here. What I usually do is adding a `--rebase` to the pull command to have it rebase instead of merge. It can be bit more work but it keeps history clean.

The PR is currently unmergeable due to this, can you fix it? One way to do it is by:

```
# replace upstream with w/e your remote is
$ git fetch upstream
$ git checkout -b 1263_bfd_disasm_v2 1263_bfd_disasm  #easy backup just in case
$ git reset upstream/master
$ git add <your changes>
$ git commit -c 30e95232830919a9ebfc94cf63fce0d4e5495c70
$ git push yourreporemotename 1263_bfd_disasm_v2:1263_bfd_disasm --force-with-lease
```

Sorry for the noise, thanks!
